flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/rules/physical/batch/BatchExecHashJoinRule.scala

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.physical.batch
+
+import org.apache.flink.table.JDouble
+import org.apache.flink.table.api.{OperatorType, PlannerConfigOptions, TableConfig, TableConfigOptions}
+import org.apache.flink.table.calcite.FlinkContext
+import org.apache.flink.table.plan.FlinkJoinRelType
+import org.apache.flink.table.plan.`trait`.FlinkRelDistribution
+import org.apache.flink.table.plan.nodes.FlinkConventions
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalJoin
+import org.apache.flink.table.plan.nodes.physical.batch.BatchExecHashJoin
+
+import org.apache.calcite.plan.RelOptRule.{any, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall, RelTraitSet}
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.rel.core.{Join, SemiJoin}
+import org.apache.calcite.util.ImmutableIntList
+
+import java.util
+
+import scala.collection.JavaConversions._
+
+/**
+  * Rule that converts [[FlinkLogicalJoin]] to [[BatchExecHashJoin]]
+  * if join keys are not empty and ShuffleHashJoin or BroadcastHashJoin are enabled.
```

Comment:
we could make this comment more explicit to replace "if join keys are not empty" to "if there exists at least one equal-join condition".